// @sap/cds-dk/lib/init/template/cloud-logging/index.js
const cds = require('../../../cds')
const { merge, sort } = require('../../merge')

const { readProject } = require('../../projectReader')
const { srv4, mtxSidecar4, approuter } = require('../../registries/mta')

module.exports = class extends require('../../plugin') {

  static hasInProduction(env) {
    return env.requires?.telemetry?.kind === 'to-cloud-logging'
  }

  options() {
    return {
      'plan': {
        type: 'string',
        help: 'Specify the service plan.',
      },
      'with-telemetry': {
        type: 'boolean',
        help: 'Add telemetry capabilities.'
      }
    }
  }

  async run() {
    const project = readProject(), { hasTelemetry, isNodejs, isJava } = project
    if (cds.cli.options['with-telemetry'] || hasTelemetry) {
      if (isNodejs) {
        project.profile ??= 'production'
        await merge(__dirname, 'files/package.json.hbs').into('package.json', { with: project })
        await sort('package.json', 'dependencies')
      } else if (isJava) {
        await merge(__dirname, 'files/java-cdsrc.json').into('.cdsrc.json')
      }
    }
  }

  async combine() {
    const project = readProject()
    const { hasMta, hasHelm, hasHelmUnifiedRuntime, hasApprouter, isJava, hasMultitenancy, srvPath } = project

    if (hasMta) {
      const cloudLogging = {
        in: 'resources',
        where: {
          type: 'org.cloudfoundry.managed-service',
          'parameters.service': 'cloud-logging'
        }
      }
      const srv = srv4(srvPath)
      const additions = [srv, cloudLogging]
      const relationships = [{
        insert: [cloudLogging, 'name'],
        into: [srv, 'requires', 'name']
      }]
      if (hasMultitenancy) {
        const mtxSidecar = mtxSidecar4(isJava ? 'mtx/sidecar' : 'gen/mtx/sidecar')
        additions.push(mtxSidecar)
        relationships.push({
          insert: [cloudLogging, 'name'],
          into: [mtxSidecar, 'requires', 'name']
        })
      }
      if (hasApprouter) {
        additions.push(approuter)
        relationships.push({
          insert: [cloudLogging, 'name'],
          into: [approuter, 'requires', 'name']
        })
      }

      project.cloudLoggingServicePlan = cds.cli.options['plan'] ?? 'standard'
      await merge(__dirname, 'files/mta.yaml.hbs').into('mta.yaml', {
        with: project, additions, relationships
      })
    }
    if (hasHelm || hasHelmUnifiedRuntime) {
      await merge(__dirname, 'files/values.yaml.hbs').into('chart/values.yaml')
    }
  }
}
