// @sap/eslint-plugin-cds/lib/rules/sql-cast-suggestion.js
'use strict'

module.exports = {
  meta: {
    schema: [{/* to avoid deprecation warning for ESLint 9 */}],
    docs: {
      description: 'Should make suggestions for possible missing SQL casts.',
      recommended: true,
      url: 'https://cap.cloud.sap/docs/tools/cds-lint/rules/sql-cast-suggestion',
    },
    type: 'suggestion',
    messages: {
      missingSQLCast: 'Potential issue - Missing SQL cast for column expression?'
    }
  },
  create: function (context) {
    return { view: checkSqlCast }

    function checkSqlCast (v) {
      // TODO: restructure and make more robust (#507)
      if (v?.query?.SET?.args) {
        for (const arg of v.query.SET.args) {
          if (arg?.SELECT) {
            // Only in UNION cases?
            for (const each of arg.SELECT.columns || []) {
              if (each) {
                const { xpr, cast } = each
                if (cast && xpr) {
                  if (!(xpr[0]?.xpr && xpr[0]?.cast)) {
                    context.report({
                      messageId: 'missingSQLCast',
                      node: context.getNode(v)
                    })
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
