// @sap/cds-dk/lib/init/template/ams/index.js
const cds = require('../../../cds')
const { readProject } = require('../../projectReader')
const { merge, sort } = require('../../merge')
const { srv4 } = require('../../registries/mta')
const mvn = require('../../mvn')

module.exports = class AmsTemplate extends require('../../plugin') {

    static help() {
        return 'authorization via AMS'
    }

    requires() {
        return ['ias']
    }

    static hasInProduction(env) {
        return env.requires?.auth?.ams || cds.cli.options?.add?.has('ams') || cds.cli.argv?.some(arg => arg?.match?.(/(^|,)ams(,|$)/))
    }

    async run() {
        const project = readProject()
        await merge(__dirname, 'files/package.json.hbs').into('package.json', { project, forceOverwrite: true })
        project.isNodejs && await sort('package.json', 'dependencies')
        await sort('package.json', 'devDependencies')
        if (project.isJava) await mvn.add('ams')
    }

    async combine() {
        const project = readProject()
        const { hasMta, hasCfManifest, hasHelm, hasHelmUnifiedRuntime, hasContainerize } = project

        if (hasMta) {
            await this.#combineWithMta(project)
        }

        if (hasCfManifest) {
            await this.#combineWithCfManifest(project)
        }

        if (hasHelm || hasHelmUnifiedRuntime) {
            await this.#combineWithHelm(project)
        }

        if (hasContainerize) {
            await this.#combineWithContainerize(project)
        }
    }

    async #combineWithMta(project) {
        const srv = srv4(project.srvPath)
        const amsPoliciesDeployer = {
            in: 'modules',
            where: { 'name': `${project.appName}-ams-policies-deployer` }
        }
        await merge(__dirname, 'files/mta.yaml.hbs').into('mta.yaml', { project, additions: [srv, amsPoliciesDeployer] })
    }

    async #combineWithCfManifest(project) {
        const amsPoliciesDeployer = {
            in: 'applications',
            where: { 'name': `${project.appName}-ams-policies-deployer` }
        }
        await merge(__dirname, 'files/manifest.yml.hbs').into('manifest.yml', { project, additions: [amsPoliciesDeployer] })
    }

    async #combineWithHelm(project) {
        const amsPoliciesDeployer = {
            in: 'dependencies',
            where: { 'alias': 'ams-policies-deployer' }
        }
        await merge(__dirname, 'files/Chart.yaml.hbs').into('chart/Chart.yaml', { project, additions: [amsPoliciesDeployer] })
        await merge(__dirname, 'files/values.yaml.hbs').into('chart/values.yaml', { project })
    }

    async #combineWithContainerize(project) {
        const amsDeployer = { in: 'modules', where: { 'name': `${project.appName}-ams-policies-deployer` } }
        await merge(__dirname, 'files/containerize.yaml.hbs').into('containerize.yaml', { project, additions: [amsDeployer] })
    }
}
